superset/db_engine_specs/questdb.py

```diff
@@ -0,0 +1,311 @@
+            parameters: BasicParametersType,
+            encrypted_extra: Optional[Dict[str, str]] = None
+    ) -> str:
+        return connection_uri(
+            parameters.get("host"),
+            int(parameters.get("port")),
+            parameters.get("username"),
+            parameters.get("password"),
+            parameters.get("database"))
+
+    @classmethod
+    def get_default_schema_for_query(cls, database, query) -> Optional[str]:
+        return 'public'
+
+    @classmethod
+    def get_text_clause(cls, clause):
+        """SQLAlchemy wrapper to ensure text clauses are escaped properly
+        :param clause: string clause with potentially unescaped characters
+        :return: text clause with escaped characters
+        """
+        if cls.allows_escaped_colons:
+            clause = clause.replace(":", "\\:")
+        return text(remove_public_schema(clause))
+
+    @classmethod
+    def get_time_grain_expressions(cls) -> Dict[Optional[str], str]:
+        """Return a dict of all supported time grains including any
+        potential added grains but excluding any potentially disabled
+        grains in the config file.
+        :return: All time grain expressions supported by the engine
+        """
+        return cls._time_grain_expressions
+
+    @classmethod
+    def epoch_to_dttm(cls) -> str:
+        """SQL expression that converts epoch (seconds) to datetime that can be used in a
+        query. The reference column should be denoted as `{col}` in the return
+        expression, e.g. "FROM_UNIXTIME({col})"
+        :return: SQL Expression
+        """
+        return '{col} * 1000000'
+
+    @classmethod
+    def convert_dttm(cls, target_type: str, dttm: datetime, *_args, **_kwargs) -> Optional[str]:
+        """Convert a Python `datetime` object to a SQL expression.
+        :param target_type: The target type of expression
+        :param dttm: The datetime object
+        :return: The SQL expression
+        """
+        type_u = target_type.upper()
+        if type_u == 'DATE':
+            return f"TO_DATE('{dttm.date().isoformat()}', 'YYYY-MM-DD')"
+        if type_u in ('DATETIME', 'TIMESTAMP'):
+            dttm_formatted = dttm.isoformat(sep=" ", timespec="microseconds")
+            return f"TO_TIMESTAMP('{dttm_formatted}', 'yyyy-MM-ddTHH:mm:ss.SSSUUUZ')"
+        return None
+
+    @classmethod
+    def get_datatype(cls, type_code: Any) -> Optional[str]:
+        """Change column type code from cursor description to string representation.
+        :param type_code: Type code from cursor description
+        :return: String representation of type code
+        """
+        return type_code.upper() if type_code and isinstance(type_code, str) else str(type_code)
+
+    @classmethod
+    def get_time_grains(cls) -> Tuple[TimeGrain, ...]:
+        """Generate a tuple of supported time grains.
+        :return: All time grains supported by the engine
+        """
+        return cls._engine_time_grains
```

Comment:
same here

Reply:
and here 9a3c588dc495ad0435666f41b455d54ee57b40fc